pkg/rook/migrate.go

```diff
@@ -0,0 +1,348 @@
+
+	osdListStrings := []string{}
+	for _, osd := range hostPathOSDs {
+		osdListStrings = append(osdListStrings, fmt.Sprintf("osd.%d", osd))
+	}
+
+	out(fmt.Sprintf("Removing hostpath OSDs %s from the cluster", strings.Join(osdListStrings, ", ")))
+	for _, osdNum := range hostPathOSDs {
+		err = safeRemoveOSD(ctx, client, osdNum)
+		if err != nil {
+			return fmt.Errorf("failed to safely remove OSD %d: %w", osdNum, err)
+		}
+	}
+
+	out("Migration completed successfully!")
+
+	return nil
+}
+
+// enableBlockDevices runs kubectl commands directly to edit the cephcluster object
+// this is because later versions of the go library do not have the relevant fields
+func enableBlockDevices(ctx context.Context, client kubernetes.Interface, cephClient *cephv1.CephV1Client) error {
+	out("Disabling directory storage and enabling block OSDs")
+	err := patchCephcluster(ctx, cephClient)
+	if err != nil {
+		return fmt.Errorf("failed to patch cephcluster to remove directory storage: %w", err)
+	}
+
+	blockDeviceOSDsExist, err := doBlockOSDsExist(ctx, client)
+	if err != nil {
+		return fmt.Errorf("failed to determine if block device OSDs existed: %w", err)
+	}
+	if !blockDeviceOSDsExist {
+		out("Restarting the rook-ceph operator to ensure modified settings take effect immediately")
+		err = restartOperator(ctx, client)
+		if err != nil {
+			return fmt.Errorf("failed to restart rook-ceph operator after patching ceph spec: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// later rook imports do not have the structs for directory storage, so any update removes it
+func patchCephcluster(ctx context.Context, cephClient *cephv1.CephV1Client) error {
+	cluster, err := cephClient.CephClusters("rook-ceph").Get(ctx, "rook-ceph", metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to get cephcluster to modify: %w", err)
+	}
+
+	truebool := true
+	cluster.Spec.Storage.UseAllDevices = &truebool
+
+	_, err = cephClient.CephClusters("rook-ceph").Update(ctx, cluster, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to update cephcluster: %w", err)
+	}
+
+	return nil
+}
+
+// checks if any block device OSDs exist - if they do, returns true
+// intended to be used to determine if the 'enable block devices' patch needs to be run
+func doBlockOSDsExist(ctx context.Context, client kubernetes.Interface) (bool, error) {
+	_, blockOSDCount, err := countRookOSDs(ctx, client)
+	if err != nil {
+		return false, fmt.Errorf("failed to find block OSD count: %w", err)
+	}
+
+	return blockOSDCount > 0, nil
+}
+
+func restartOperator(ctx context.Context, client kubernetes.Interface) error {
+	err := client.CoreV1().Pods("rook-ceph").DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "app=rook-ceph-operator"})
+	if err != nil {
+		return fmt.Errorf("failed to delete rook-ceph-operator pods: %w", err)
+	}
+
+	return nil
+}
+
+func safeRemoveOSD(ctx context.Context, client kubernetes.Interface, osdNum int64) error {
```

Comment:
Small point, could be categorized as a nit though.  In this function we reweight to 0 - however, that doesn't necessarily clear all the data ready for the removal of the OSD.  I would usually want to use `ceph osd out osd.%d` in this case, to ensure that Ceph rebalances all the data from the OSD and ejects it (as if the disk was dead).  Once the OSD is out, and Ceph is healthy, you can be confident that removal is safe, but reweighting does not give you that guarantee as it merely adjusts the crush map.  I see you do wait for Ceph to settle after scaling down the deployment, so the operation as is works fine in most cases, however if you find that Ceph is not healthy after marking the OSD out you have a slightly harder rollback than if you simply marked it out rather than removing the pod.  This is only an issue on a nearly-full cluster, particularly if the block devices are too small, or if there's too few ODSs to match crush rules.

Reply:
(I've also made sure that it's running `ceph osd safe-to-destroy osd.%d` before actually deleting things)